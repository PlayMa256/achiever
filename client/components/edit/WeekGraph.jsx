import React from 'react';
import PropTypes from 'prop-types';
import TimeDuration from 'time-duration';

import GaugeBar from '../ui/GaugeBar';

import { timesAreValid, calculateLabouredHours } from '../../utils';

import '../../styles/weekGraph.styl';

const renderDayBar = (dailyContractedHours, controlDate, storedTimes) => (dayInfo) => {
	const current = (dayInfo.date === controlDate.format('YYYY-MM-DD')) ?
		(timesAreValid(storedTimes) &&
			new TimeDuration(calculateLabouredHours(storedTimes)).toMinutes()
		) || 0 :
		new TimeDuration(dayInfo.total).toMinutes();
	const reference = new TimeDuration(dailyContractedHours).toMinutes();

	return (
		<GaugeBar
			key={dayInfo.date}
			currentValue={current}
			referenceValue={reference}
			verticalBar
		/>
	);
};

const WeekGraph = (props) => {
	const {
		dailyContractedHours,
		weekEntries,
		controlDate,
		storedTimes
	} = props;

	if (weekEntries.timeEntries && weekEntries.timeEntries.length &&
		weekEntries.timeEntries.length === 7) {
		const entriesForWeekDays = weekEntries.timeEntries
			.filter((dayEntry, index) => index !== 0 && index !== 6);
		return (
			<div className="week-graph">
				{entriesForWeekDays.map(renderDayBar(dailyContractedHours, controlDate, storedTimes))}
			</div>
		);
	}
	return null;
};

WeekGraph.propTypes = {
	weekEntries: PropTypes.object,
	dailyContractedHours: PropTypes.string,
	controlDate: PropTypes.object,
	storedTimes: PropTypes.array
};

WeekGraph.defaultProps = {
	weekEntries: { timeEntries: null },
	dailyContractedHours: '8:00',
	controlDate: {},
	storedTimes: [{}, {}, {}, {}]
};

export default WeekGraph;
