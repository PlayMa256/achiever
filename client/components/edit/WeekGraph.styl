@import "../../styles/_colours.styl"
@import "../../styles/_standards.styl"

.week-graph
    width: 245px
    padding: 9px 10px

    .gauge-bar:not(:last-child)
        margin-bottom: 5px