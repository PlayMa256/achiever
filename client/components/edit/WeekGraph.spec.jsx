import React from 'react';
import { shallow } from 'enzyme';

import moment from 'moment';

import WeekGraph from './WeekGraph';

Date.now = jest.fn(() => new Date(Date.UTC(2018, 1, 14, 12)).valueOf());

const defaultProps = {
	dailyContractedHours: '8:00',
	controlDate: moment('2018-02-14'),
	weekEntries: {
		timeEntries: [
			{
				date: '2018-02-11',
				phase: 'Project Phase Name',
				activity: 'Code/ Document Reviews',
				startTime: '9:28',
				startBreakTime: '11:35',
				endBreakTime: '12:35',
				endTime: '18:28',
				total: '8:00',
				__typename: 'TimeEntry'
			},
			{
				date: '2018-02-12',
				phase: 'Project Phase Name',
				activity: 'Code/ Document Reviews',
				startTime: '9:28',
				startBreakTime: '11:35',
				endBreakTime: '12:35',
				endTime: '18:28',
				total: '8:00',
				__typename: 'TimeEntry'
			},
			{
				date: '2018-02-13',
				phase: 'Project Phase Name',
				activity: 'Code/ Document Reviews',
				startTime: '9:28',
				startBreakTime: '11:35',
				endBreakTime: '12:35',
				endTime: '18:28',
				total: '8:00',
				__typename: 'TimeEntry'
			},
			{
				date: '2018-02-14',
				phase: 'Project Phase Name',
				activity: 'Code/ Document Reviews',
				startTime: '',
				startBreakTime: '',
				endBreakTime: '',
				endTime: '',
				total: '8:00',
				__typename: 'TimeEntry'
			},
			{
				date: '2018-02-15',
				phase: 'Project Phase Name',
				activity: 'Code/ Document Reviews',
				startTime: '9:28',
				startBreakTime: '11:35',
				endBreakTime: '12:35',
				endTime: '18:28',
				total: '8:00',
				__typename: 'TimeEntry'
			},
			{
				date: '2018-02-16',
				phase: 'Project Phase Name',
				activity: 'Code/ Document Reviews',
				startTime: '9:28',
				startBreakTime: '',
				endBreakTime: '',
				endTime: '18:28',
				total: '8:00',
				__typename: 'TimeEntry'
			},
			{
				date: '2018-02-17',
				phase: 'Project Phase Name',
				activity: 'Code/ Document Reviews',
				startTime: '9:28',
				startBreakTime: '11:35',
				endBreakTime: '12:35',
				endTime: '18:28',
				total: '8:00',
				__typename: 'TimeEntry'
			}
		],
		total: '56:00'
	},
	storedTimes: [
		{ hours: 7, minutes: 0 },
		{ hours: 12, minutes: 0 },
		{ hours: 0, minutes: 0 },
		{ hours: 0, minutes: 0 }
	]
};

describe('WeekGraph', () => {
	describe('render', () => {
		it('should render nothing using min props', () => {
			const wrapper = shallow(<WeekGraph />);
			expect(wrapper).toMatchSnapshot();
		});
		it('should render GaugeBars if props are correct', () => {
			const wrapper = shallow(<WeekGraph
				weekEntries={defaultProps.weekEntries}
				dailyContractedHours={defaultProps.dailyContractedHours}
				controlDate={defaultProps.controlDate}
				storedTimes={defaultProps.storedTimes}
			/>);
			expect(wrapper).toMatchSnapshot();
		});
	});
});
