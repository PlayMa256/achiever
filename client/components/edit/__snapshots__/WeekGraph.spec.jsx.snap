// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeekGraph render should render GaugeBars if props are correct 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WeekGraph
    controlDate={"2018-02-14T00:00:00.000Z"}
    dailyContractedHours="8:00"
    storedTimes={
      Array [
        Object {
          "hours": 7,
          "minutes": 0,
        },
        Object {
          "hours": 12,
          "minutes": 0,
        },
        Object {
          "hours": 0,
          "minutes": 0,
        },
        Object {
          "hours": 0,
          "minutes": 0,
        },
      ]
    }
    weekEntries={
      Object {
        "timeEntries": Array [
          Object {
            "__typename": "TimeEntry",
            "activity": "Code/ Document Reviews",
            "date": "2018-02-11",
            "endBreakTime": "12:35",
            "endTime": "18:28",
            "phase": "Project Phase Name",
            "startBreakTime": "11:35",
            "startTime": "9:28",
            "total": "8:00",
          },
          Object {
            "__typename": "TimeEntry",
            "activity": "Code/ Document Reviews",
            "date": "2018-02-12",
            "endBreakTime": "12:35",
            "endTime": "18:28",
            "phase": "Project Phase Name",
            "startBreakTime": "11:35",
            "startTime": "9:28",
            "total": "8:00",
          },
          Object {
            "__typename": "TimeEntry",
            "activity": "Code/ Document Reviews",
            "date": "2018-02-13",
            "endBreakTime": "12:35",
            "endTime": "18:28",
            "phase": "Project Phase Name",
            "startBreakTime": "11:35",
            "startTime": "9:28",
            "total": "8:00",
          },
          Object {
            "__typename": "TimeEntry",
            "activity": "Code/ Document Reviews",
            "date": "2018-02-14",
            "endBreakTime": "",
            "endTime": "",
            "phase": "Project Phase Name",
            "startBreakTime": "",
            "startTime": "",
            "total": "8:00",
          },
          Object {
            "__typename": "TimeEntry",
            "activity": "Code/ Document Reviews",
            "date": "2018-02-15",
            "endBreakTime": "12:35",
            "endTime": "18:28",
            "phase": "Project Phase Name",
            "startBreakTime": "11:35",
            "startTime": "9:28",
            "total": "8:00",
          },
          Object {
            "__typename": "TimeEntry",
            "activity": "Code/ Document Reviews",
            "date": "2018-02-16",
            "endBreakTime": "",
            "endTime": "18:28",
            "phase": "Project Phase Name",
            "startBreakTime": "",
            "startTime": "9:28",
            "total": "8:00",
          },
          Object {
            "__typename": "TimeEntry",
            "activity": "Code/ Document Reviews",
            "date": "2018-02-17",
            "endBreakTime": "12:35",
            "endTime": "18:28",
            "phase": "Project Phase Name",
            "startBreakTime": "11:35",
            "startTime": "9:28",
            "total": "8:00",
          },
        ],
        "total": "56:00",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <GaugeBar
          currentValue={480}
          referenceValue={480}
          verticalBar={true}
        />,
        <GaugeBar
          currentValue={0}
          referenceValue={480}
          verticalBar={true}
        />,
        <GaugeBar
          currentValue={480}
          referenceValue={480}
          verticalBar={true}
        />,
        <GaugeBar
          currentValue={480}
          referenceValue={480}
          verticalBar={true}
        />,
        <GaugeBar
          currentValue={480}
          referenceValue={480}
          verticalBar={true}
        />,
      ],
      "className": "week-graph",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "2018-02-12",
        "nodeType": "function",
        "props": Object {
          "currentValue": 480,
          "referenceValue": 480,
          "verticalBar": true,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2018-02-13",
        "nodeType": "function",
        "props": Object {
          "currentValue": 0,
          "referenceValue": 480,
          "verticalBar": true,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2018-02-14",
        "nodeType": "function",
        "props": Object {
          "currentValue": 480,
          "referenceValue": 480,
          "verticalBar": true,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2018-02-15",
        "nodeType": "function",
        "props": Object {
          "currentValue": 480,
          "referenceValue": 480,
          "verticalBar": true,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2018-02-16",
        "nodeType": "function",
        "props": Object {
          "currentValue": 480,
          "referenceValue": 480,
          "verticalBar": true,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <GaugeBar
            currentValue={480}
            referenceValue={480}
            verticalBar={true}
          />,
          <GaugeBar
            currentValue={0}
            referenceValue={480}
            verticalBar={true}
          />,
          <GaugeBar
            currentValue={480}
            referenceValue={480}
            verticalBar={true}
          />,
          <GaugeBar
            currentValue={480}
            referenceValue={480}
            verticalBar={true}
          />,
          <GaugeBar
            currentValue={480}
            referenceValue={480}
            verticalBar={true}
          />,
        ],
        "className": "week-graph",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "2018-02-12",
          "nodeType": "function",
          "props": Object {
            "currentValue": 480,
            "referenceValue": 480,
            "verticalBar": true,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2018-02-13",
          "nodeType": "function",
          "props": Object {
            "currentValue": 0,
            "referenceValue": 480,
            "verticalBar": true,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2018-02-14",
          "nodeType": "function",
          "props": Object {
            "currentValue": 480,
            "referenceValue": 480,
            "verticalBar": true,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2018-02-15",
          "nodeType": "function",
          "props": Object {
            "currentValue": 480,
            "referenceValue": 480,
            "verticalBar": true,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2018-02-16",
          "nodeType": "function",
          "props": Object {
            "currentValue": 480,
            "referenceValue": 480,
            "verticalBar": true,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`WeekGraph render should render nothing using min props 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WeekGraph
    controlDate={Object {}}
    dailyContractedHours="8:00"
    storedTimes={
      Array [
        Object {},
        Object {},
        Object {},
        Object {},
      ]
    }
    weekEntries={
      Object {
        "timeEntries": null,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
